lib/gears/surface.lua

```diff
@@ -205,6 +206,52 @@ function surface.apply_shape_bounding(draw, shape, ...)
   draw.shape_bounding = img._native
 end
 
+local function run_in_hierarchy(self, cr, width, height)
+    local function main_layout(h)
+        h:update({dpi=96}, self, width, height)
+    end
+
+    local function redraw(h)
+        h:draw({dpi=96}, cr)
+    end
+
+    local h = hierarchy.new({dpi=96}, self, width, height, redraw, main_layout, {})
+
+    main_layout(h)
+    h:_redraw()
+
+    return h
+end
+
+--- Create an SVG file with this widget content.
+-- This is dynamic, so the SVG will be updated along with the widget content.
+-- because of this, the painting may happen hover multiple event loop cycles.
+-- @tparam widget widget A widget
+-- @tparam string path The output file path
+-- @tparam number width The surface width
+-- @tparam number height The surface height
+function surface.widget_to_svg(widget, path, width, height)
+    local img = cairo.SvgSurface.create(path, width, height)
+    local cr = cairo.Context(img)
+
+    return run_in_hierarchy(widget, cr, width, height)
```

Comment:
Replace this with `hierarchy.new({dpi=96}, widget, width, height, function() end, function() end, {}):draw({dpi=96}, cr)` (for extra niceness, you can make and use a `local function no_op() end` and move the context into a variable, too).

There will be no relayouts necessary here and no redraws either (in fact, redraws would be bad, because the widget would be drawn to the SVG twice).

(Same thing below)